src/plugins/asp/aspect/clingo_access.cpp

```diff
@@ -0,0 +1,617 @@
+ * @brief The mutex to protect the callback vectors.
+ *
+ * @property ClingoAccess::model_callbacks_
+ * @brief The functions to call on a new model.
+ *
+ * @property ClingoAccess::finish_callbacks_
+ * @brief The functions to call, when the solving process finished.
+ *
+ * @property ClingoAccess::ground_callback_
+ * @brief The callback for the grounding.
+ *
+ * @property ClingoAccess::async_handle_
+ * @brief The handle for the async solving.
+ *
+ * @property ClingoAccess::debug_level_
+ * @brief Which debug outputs should be printed.
+ */
+
+/**
+ * @brief Extracts the symbols from the new model and calls the registered functions.
+ * @param[in] model The new model.
+ * @return If the solving process should compute more models (if there are any).
+ */
+bool
+ClingoAccess::on_model(Clingo::Model& model)
+{
+	MutexLocker locker1(&model_mutex_);
+	model_symbols_ = model.symbols(debug_level_ >= AllModelSymbols ? Clingo::ShowType::All : Clingo::ShowType::Shown);
+
+	if (debug_level_ >= Time)	{
+		logger_->log_info(log_comp_.c_str(), "New %smodel found: #%d",
+		                  model.optimality_proven() ? "optimal " : "",
+		                  ++model_counter_);
+
+		if (debug_level_ >= Models)	{
+			/* To save (de-)allocations just move found symbols at the end
+			 * of the vector and move the end iterator to the front. After
+			 * this everything in [begin, end) is in oldSymbols but not in
+			 * symbols. */
+			auto begin = old_symbols_.begin(), end = old_symbols_.end();
+
+			for (const Clingo::Symbol& symbol : model_symbols_)	{
+				auto iter = std::find(begin, end, symbol);
+				if (iter == end) {
+					logger_->log_info(log_comp_.c_str(), "New Symbol: %s", symbol.to_string().c_str());
+				} else {
+					std::swap(*iter, *--end);
+				}
+			}
+
+			for (; begin != end; ++begin) {
+				logger_->log_info(log_comp_.c_str(), "Symbol removed: %s", begin->to_string().c_str());
+			}
+		}
+	}
+
+	old_symbols_ = model_symbols_;
+
+	bool ret = false;
+
+	MutexLocker locker2(&callback_mutex_);
+	for (const auto& cb : model_callbacks_) {
+		ret |= (*cb)();
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Calls the callbacks for the end of the solving.
+ * @param[in] result The result of the solving process.
+ */
+void
+ClingoAccess::on_finish(Clingo::SolveResult result)
+{
+	if (debug_level_ >= Time) {
+		logger_->log_info(log_comp_.c_str(), "Solving nearly done.");
+	}
+
+	BoolMutexLocker locker1(&control_mutex_, control_is_locked_);
+	MutexLocker locker2(&callback_mutex_);
+	for (const auto& cb : finish_callbacks_) {
+		(*cb)(result);
+	}
+
+	std::thread blocking_thread(
+	  [this](void)
+	  {
+		  async_handle_.wait();
+		  if (debug_level_ >= Time)	{
+				logger_->log_info(log_comp_.c_str(), "Solving done.");
+			}
+			solving_ = false;
+			return;
+	  });
+	blocking_thread.detach();
+	return;
+}
+
+/**
+ * @brief Allocates the control object and initializes the logger.
+ */
+void
+ClingoAccess::alloc_control()
+{
+	assert(!control_);
```

Comment:
We could also throw an exception here

Reply:
It is not something that the run-time execution code path could do. It really is something that is useful during development when bugs in this very code were a concern, but the situations for 'control_' and 'alloc_control()' are very constrained. It happens once in the ctor, where the member is in the initializer list, and in reset, where resetting and re-creating are done together. So I think an exception in this case here is not necessary.